src/net_processing.cpp

```diff
@@ -1014,32 +1039,45 @@ void PeerManagerImpl::FindNextBlocksToDownload(NodeId nodeid, unsigned int count
     // Make sure pindexBestKnownBlock is up to date, we'll need it.
     ProcessBlockAvailability(nodeid);
 
-    if (state->pindexBestKnownBlock == nullptr || state->pindexBestKnownBlock->nChainWork < m_chainman.ActiveChain().Tip()->nChainWork || state->pindexBestKnownBlock->nChainWork < nMinimumChainWork) {
+    const CChain& our_chain = chainstate->m_chain;
+    const CBlockIndex* our_tip = our_chain.Tip();
+
+    if (state->pindexBestKnownBlock == nullptr
+            || state->pindexBestKnownBlock->nChainWork < our_tip->nChainWork
+            || state->pindexBestKnownBlock->nChainWork < nMinimumChainWork) {
         // This peer has nothing interesting.
         return;
     }
 
-    if (state->pindexLastCommonBlock == nullptr) {
+    if (!state->chainstate_to_last_common_block.count(chainstate)) {
         // Bootstrap quickly by guessing a parent of our best tip is the forking point.
         // Guessing wrong in either direction is not a problem.
-        state->pindexLastCommonBlock = m_chainman.ActiveChain()[std::min(state->pindexBestKnownBlock->nHeight, m_chainman.ActiveChain().Height())];
+        //
+        // Namespace this by chainstate so that we can simultaneously sync two
+        // separate chainstates at different heights.
+        state->chainstate_to_last_common_block[chainstate] = our_chain[
```

Comment:
0ef00576 maybe safer on first use in this function to raise instead of failing silently/UB
```suggestion
        state->chainstate_to_last_common_block.at(chainstate) = our_chain[
```

Reply:
`x[k]` for a map [isn't undefined behaviour](https://en.cppreference.com/w/cpp/container/map/operator_at), it creates the entry (setting it to nullptr here), then the `=` replaces it. `insert_or_assign(chainstate, our_chain[blah])` would arguably be better I guess.